### PR TITLE
(Mainly) Sonar tweaks and fixes

### PR DESCRIPTION
**Pull request by SuicidalInsanity**
- Add missing sonar checks for AIs trying to fire torpedoes.
- Fix AI not turning on Active sonar when trying to fire INS torpedoes if no passive sonar on vessel.
- AI can now make use of non-DynamicRadar radars for missile targeting when being actively targeted by incoming anti-radiation ordinance.
- AI visual detection of incoming missiles within visual range no longer requires a RWR. RWR detection of incoming missiles still requires a RWR on the craft.
- Tweak AI detection and reaction to incoming GPS/INS missiles, they should detect and react to them a bit more consistently now
- Competition UI text in the upper left corner now properly scales when adjusting UI_SCALE.
- Add ability to set teams' Allies to the Team Selector GUI.
- Fix craft equipped with repulsors breaking ground spawn.
- Acoustic Decoys will no longer defy gravity and fly if launched out of water.
- 'Infinite Ordinance' now applies to Countermeasure droppers.
- Adjust cmThreshold timing for SLW countermeasures to 5x time set in WM to account for torpedoes being substantially slower than missiles. This should allow proper CM behavior against both torpedoes and missiles without compromising one or the other.
- Fix Electrolaser NRE when firing on a target vessel.
- Tweak acoustic propagation loss rate for passive acoustic torpedoes, should have a bit further effective detection range now.
- Fix INS guidance ordinance not configured with a datalink immediately triggering the lost lock condition and self-destructing.
- Passive Sonar torpedoes now home in on loudest part, not CoM (unless targetCoM set in missile c.fg)
- Fix Terminal Guidance Passive Sonar torpedoes not entering terminal guidance.

### CODE REVIEW

File: BDArmory/UI/BDTeamSelector.cs [hunk 1]
```diff
@@ -49,8 +49,14 @@ void SetVisible(bool visible)
         private void TeamSelectorWindow(int id)
         {
             height = margin;
+            GUI.Label(new Rect(margin, height, width - 2 * margin, buttonHeight), StringUtils.Localize("#LOC_BDArmory_SelectTeam"), BDArmorySetup.BDGuiSkin.label);
```
Comment (BrettRyland) [inline BDArmory/UI/BDTeamSelector.cs:52]: Maybe add a drag region and hide the window when the VS gets hidden? Also, the window ID should be unique. This diff should cover these, fix a spelling mistake and also make the allies section auto-adjust to the correct height:
```diff
diff --git a/BDArmory/UI/BDTeamSelector.cs b/BDArmory/UI/BDTeamSelector.cs
index 88f4b8ae3..9e283e282 100644
--- a/BDArmory/UI/BDTeamSelector.cs
+++ b/BDArmory/UI/BDTeamSelector.cs
@@ -19,3 +19,3 @@ public class BDTeamSelector : MonoBehaviour
         const float buttonGap = 2;
-        const float newTeanButtonWidth = 40;
+        const float newTeamButtonWidth = 40;
         const float scrollWidth = 20;
@@ -29,2 +29,4 @@ public class BDTeamSelector : MonoBehaviour
         private Vector2 scrollPosition = Vector2.zero;
+        Rect alliesRect;
+        GUIStyle alliesStyle;
 
@@ -36,3 +38,2 @@ public void Open(MissileFire weaponManager, Vector2 position)
         {
-            SetVisible(true);
             targetWeaponManager = weaponManager;
@@ -40,5 +41,6 @@ public void Open(MissileFire weaponManager, Vector2 position)
             windowLocation = position;
+            SetVisible(true);
         }
 
-        void SetVisible(bool visible)
+        public void SetVisible(bool visible)
         {
@@ -46,2 +48,12 @@ void SetVisible(bool visible)
             GUIUtils.SetGUIRectVisible(guiCheckIndex, visible);
+            if (visible)
+            {
+                window = new Rect(
+                    Mathf.Min(windowLocation.x, Screen.width - (scrollable ? width + scrollWidth : width)),
+                    Mathf.Min(windowLocation.y, Screen.height - height),
+                    width,
+                    scrollable ? Screen.height / 2 + buttonHeight + buttonGap + 2 * margin : height
+                );
+                alliesStyle ??= new GUIStyle(BDArmorySetup.BDGuiSkin.label) { wordWrap = true, alignment = TextAnchor.UpperLeft, fixedWidth = width };
+            }
         }
@@ -52,2 +64,3 @@ private void TeamSelectorWindow(int id)
             GUI.Label(new Rect(margin, height, width - 2 * margin, buttonHeight), StringUtils.Localize("#LOC_BDArmory_SelectTeam"), BDArmorySetup.BDGuiSkin.label);
+            GUI.DragWindow(new Rect(margin, margin, width - 2 * margin - buttonHeight, buttonHeight));
             if (GUI.Button(new Rect(width - 18, 2, 18, 18), "X"))
@@ -58,6 +71,6 @@ private void TeamSelectorWindow(int id)
             // Team input field
-            newTeamName = GUI.TextField(new Rect(margin, height, width - buttonGap - 2 * margin - newTeanButtonWidth, buttonHeight), newTeamName, 30);
+            newTeamName = GUI.TextField(new Rect(margin, height, width - buttonGap - 2 * margin - newTeamButtonWidth, buttonHeight), newTeamName, 30);
 
             // New team button
-            Rect newTeamButtonRect = new Rect(width - margin - newTeanButtonWidth, height, newTeanButtonWidth, buttonHeight);
+            Rect newTeamButtonRect = new Rect(width - margin - newTeamButtonWidth, height, newTeamButtonWidth, buttonHeight);
             if (GUI.Button(newTeamButtonRect, StringUtils.Localize("#LOC_BDArmory_Generic_New"), BDArmorySetup.BDGuiSkin.button))//"New"
@@ -127,4 +140,6 @@ private void TeamSelectorWindow(int id)
             string allies = string.Join("; ", targetWeaponManager.Team.Allies);
-            GUI.Label(new Rect(margin, height, width - 2 * margin, buttonHeight), allies);
-            height += buttonHeight + buttonGap * Mathf.Clamp(Mathf.CeilToInt(allies.Length / 50), 1, 4);
+            alliesRect = new Rect(margin, height, width - 2 * margin, buttonHeight)
+            { height = alliesStyle.CalcHeight(new GUIContent(allies), width) };
+            GUI.Label(alliesRect, allies, alliesStyle);
+            height += alliesRect.height;
             if (scrollable)
@@ -147,2 +162,3 @@ private void TeamSelectorWindow(int id)
             height += margin;
+            window.height = scrollable ? Screen.height / 2 + buttonHeight + buttonGap + 2 * margin : height;
             GUIUtils.RepositionWindow(ref window);
@@ -164,9 +180,4 @@ protected virtual void OnGUI()
                 {
-                    var clientRect = new Rect(
-                        Mathf.Min(windowLocation.x, Screen.width - (scrollable ? width + scrollWidth : width)),
-                        Mathf.Min(windowLocation.y, Screen.height - height),
-                        width,
-                        scrollable ? Screen.height / 2 + buttonHeight + buttonGap + 2 * margin : height);
-                    if (BDArmorySettings.UI_SCALE != 1) GUIUtility.ScaleAroundPivot(BDArmorySettings.UI_SCALE * Vector2.one, clientRect.position);
-                    window = GUI.Window(10591029, clientRect, TeamSelectorWindow, "", BDArmorySetup.BDGuiSkin.window);
+                    if (BDArmorySettings.UI_SCALE != 1) GUIUtility.ScaleAroundPivot(BDArmorySettings.UI_SCALE * Vector2.one, window.position);
+                    window = GUI.Window(GUIUtility.GetControlID(FocusType.Passive), window, TeamSelectorWindow, "", BDArmorySetup.BDGuiSkin.window);
                     GUIUtils.UpdateGUIRect(window, guiCheckIndex);
diff --git a/BDArmory/UI/LoadedVesselSwitcher.cs b/BDArmory/UI/LoadedVesselSwitcher.cs
index 884ba9154..036cabb59 100644
--- a/BDArmory/UI/LoadedVesselSwitcher.cs
+++ b/BDArmory/UI/LoadedVesselSwitcher.cs
@@ -323,2 +323,3 @@ public void SetVisible(bool visible)
             GUIUtils.SetGUIRectVisible(_guiCheckIndex, visible);
+            if (!visible && BDTeamSelector.Instance) BDTeamSelector.Instance.SetVisible(false);
         }
@@ -896,3 +897,3 @@ void AddVesselSwitcherWindowEntry(MissileFire wm, string team, float height, flo
                 {
-                    BDTeamSelector.Instance.Open(wm, new Vector2(Input.mousePosition.x, Screen.height - Input.mousePosition.y));
+                    BDTeamSelector.Instance.Open(wm, new Vector2(Input.mousePosition.x + 2 * _buttonHeight + _margin, Screen.height - Input.mousePosition.y));
                 }
```

File: BDArmory/Control/MissileFire.cs [hunk 2]
```diff
@@ -6481,30 +6494,33 @@ bool CheckEngagementEnvelope(IBDWeapon weaponCandidate, float distanceToTarget,
...
                             {
-                                if (!results.foundAntiRadiationMissile)
+                                if (!results.foundAntiRadiationMissile || !DynamicRadarOverride)
                                 {
                                     using (List<ModuleRadar>.Enumerator rd = radars.GetEnumerator())
                                         while (rd.MoveNext())
                                         {
                                             if (rd.Current != null && rd.Current.sonarMode == ModuleRadar.SonarModes.None)
                                             {
+                                                if (rd.Current.DynamicRadar && results.foundAntiRadiationMissile) continue; //don't enable radar if incoming HARM, unless radar is specifically set to to be used regardless
```
Comment (BrettRyland) [inline BDArmory/Control/MissileFire.cs:6518]: In the comment: "to to" -> "to"

File: BDArmory/Competition/BDACompetitionMode.cs [hunk 3]
```diff
@@ -18,6 +18,7 @@
 using BDArmory.VesselSpawning;
 using BDArmory.Weapons.Missiles;
 using BDArmory.Weapons;
+using UnityEngine.UIElements;
```
[the remark this answers is not in the document]
Comment (SuicidalInsanity) [reply BDArmory/Competition/BDACompetitionMode.cs:21]: Thanks, Visual Studio.

File: BDArmory/GameModes/ModuleSpaceFriction.cs [hunk 4]
```diff
@@ -221,33 +240,38 @@ public void FixedUpdate()
             }
             if (this.part.vessel.situation != Vessel.Situations.ORBITING || this.part.vessel.situation != Vessel.Situations.DOCKED || this.part.vessel.situation != Vessel.Situations.ESCAPING || this.part.vessel.situation != Vessel.Situations.PRELAUNCH)
             {
-                if (BDArmorySettings.SF_REPULSOR || RepulsorOverride)
+                if (BDArmorySettings.SF_REPULSOR || (RepulsorOverride && repulsorActivated))
```
[the remark this answers is not in the document]
Comment (SuicidalInsanity) [reply BDArmory/GameModes/ModuleSpaceFriction.cs:243]: Hrm. Probably, if someone was using the original hack that turned landing gear into repulsors and wanted to do ground spawn.

File: BDArmory/Radar/RadarUtils.cs [hunk 5]
```diff
@@ -1949,7 +1950,7 @@ public static ViewScanResults GuardScanInDirection(MissileFire myWpnManager, Tra
                     float vesselDistanceSqr = (loadedvessels.Current.transform.position - position).sqrMagnitude;
                     //BDATargetManager.ClearRadarReport(loadedvessels.Current, myWpnManager); //reset radar contact status
                     if (vesselDistanceSqr < maxRWRDistance * maxRWRDistance && Vector3.Angle(vesselProjectedDirection, lookDirection) < fov / 2f) // && Vector3.Angle(loadedvessels.Current.transform.position - position, -myWpnManager.transform.forward) < myWpnManager.guardAngle / 2f) //WM facing direction? that s going to cause issues for any that aren't mounted pointing forward if guardAngle < 360; check combatSeat forward vector
-                    {
+                    { //wait, RWR detection is also clamped to visual FOV? Should it be clamped to visual FOV?
```
[the remark this answers is not in the document]
Comment (SuicidalInsanity) [reply BDArmory/Radar/RadarUtils.cs:1953]: In theory it should be clamped to whatever FoV the Radar Warning Receiver mounted on the plane possessed.
Would be easy enough to add a rwrFOV field to The RWR, and update the WeaponManager.cfg appropriately

File: BDArmory/UI/BDATargetManager.cs [hunk 6]
```diff
@@ -826,17 +831,18 @@ public static TargetSignatureData GetAcousticTarget(Vessel sourceVessel, Vessel
...
                     else // Otherwise, pick the highest noise score
                     {
                         if (score > finalScore)
                         {
                             finalScore = score;
-                            finalData = new TargetSignatureData(vessel, score);
+                            finalData = new TargetSignatureData(vessel, score, targetCoM ? null : AcousticSig.Item2);
                         }
                     }
+                    if (BDArmorySettings.DEBUG_RADAR) Debug.Log($"[GetAcousticTarget] soundScore of {vessel.GetName()} at angle {angle}° is {score}");
```
[the remark this answers is not in the document]
Comment (SuicidalInsanity) [reply BDArmory/UI/BDATargetManager.cs:845]: Yes. Sorry, I always seem to miss one debug message somewhere.

**Pull request merged**